flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/source/reader/AvroGenericRecordReaderFunction.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.reader;
+
+import org.apache.avro.generic.GenericRecord;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.flink.source.AvroGenericRecordFileScanTaskReader;
+import org.apache.iceberg.flink.source.DataIterator;
+import org.apache.iceberg.flink.source.RowDataFileScanTaskReader;
+import org.apache.iceberg.flink.source.RowDataToAvroGenericRecordConverter;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/** Read Iceberg rows as {@link GenericRecord}. */
+public class AvroGenericRecordReaderFunction extends DataIteratorReaderFunction<GenericRecord> {
+  private final String tableName;
+  private final Schema readSchema;
+  private final FileIO io;
+  private final EncryptionManager encryption;
+  private final RowDataFileScanTaskReader rowDataReader;
+
+  private transient RowDataToAvroGenericRecordConverter converter;
+
+  /**
+   * Create a reader function without projection and name mapping. Column name is case-insensitive.
+   */
+  public static AvroGenericRecordReaderFunction fromTable(Table table) {
+    return new AvroGenericRecordReaderFunction(
+        table.name(),
+        new Configuration(),
+        table.schema(),
+        null,
+        null,
+        false,
```

Comment:
Could we get `case-sensitive` from `FlinkReadOptions`? Or provide `fromTable(Table table, boolean caseSensitive)?

Reply:
``fromTable(Table table, boolean caseSensitive)` is not working to work, because we also have totally three read configs.
```
      Schema projectedSchema,
      String nameMapping,
      boolean caseSensitive,
```

Regarding using `FlinkReadOptions`, are you suggesting another factory method like this? We need all three to construct the `FlinkReadOptions`.
```
AvroGenericRecordReaderFunction fromTableAndReadConfig(Table table, Map<String, String> readOptions, ReadableConfig readableConfig)
```

Reply:
I suggest the following method because my personal feeling is that `caseSensitive` is used a little more frequently. Of course, what you said is also reasonable, it is ok for me.
```
  public static AvroGenericRecordReaderFunction fromTable(Table table, boolean caseSensitive) {
    return new AvroGenericRecordReaderFunction(
        table.name(),
        new Configuration(),
        table.schema(),
        null,
        null,
        caseSensitive,
        table.io(),
        table.encryption());
  }
```
Regarding using `FlinkReadOptions`. Essentially, I'm interested in whether it can be built internally, rather than by the user.
So far, we've had three ReaderFuntion: `AvroGenericRecordReaderFunction` \ `MetaDataReaderFunction` \ `RowDataReaderFunction`, maybe we'll have new members in the future. Could we provide an enum for the user to choose from and then help build it automatically in the `IcebergSourceBuilder` instead of letting the user create it themselves?

Reply:
> caseSensitive is used a little more frequently.

I am not sure about that. I would think projection might be common too. name mapping might also be useful if the Avro schema is not identical (e.g. same fields and order), although we don't support name mapping right now for Avro records.

>  Could we provide an enum for the user to choose from and then help build it automatically in the IcebergSourceBuilder instead of letting the user create it themselves?

I agree with that point. By default, users shouldn't need to construct the `RowDataReaderFunction` as `IcebergSource` builds it internally if not set. We provide the `ReaderFunction` to `IcebergSource#Builder` to support  more customized reader like this. As for `BaseMetadataTable`, users probably only indirectly use via Flink SQL for inspecting metadata tables, right? I doubt users would ever need to construct it programmatically.

```
      if (readerFunction == null) {
        if (table instanceof BaseMetadataTable) {
          MetaDataReaderFunction rowDataReaderFunction =
              new MetaDataReaderFunction(
                  flinkConfig, table.schema(), context.project(), table.io(), table.encryption());
          this.readerFunction = (ReaderFunction<T>) rowDataReaderFunction;
        } else {
          RowDataReaderFunction rowDataReaderFunction =
              new RowDataReaderFunction(
                  flinkConfig,
                  table.schema(),
                  context.project(),
                  context.nameMapping(),
                  context.caseSensitive(),
                  table.io(),
                  table.encryption());
          this.readerFunction = (ReaderFunction<T>) rowDataReaderFunction;
        }
```

Reply:
Agree with you. Can we provide a similar entry where the user simply tells the Builder that he wants to use `AvroGenericRecordReaderFunction` , and then lets the Builder create it?

> users probably only indirectly use via Flink SQL for inspecting metadata tables, right? I doubt users would ever need to construct it programmatically.

Yes, that is right. However, you do remind me that we should also allow users to query metadata tables in the flink java api, right?

Reply:
> Can we provide a similar entry where the user simply tells the Builder that he wants to use AvroGenericRecordReaderFunction , and then lets the Builder create it?

I see your point. I agree that it is cumbersome for users to construct `AvroGenericRecordReaderFunction` or `RowDataReaderFunction`. Since we can extract type information from generic `T`, one idea would to add a setter for type class to the builder. If the type class is `RowData`, builder constructs `RowDataReaderFunction`. If it is Avro `GenericRecord, builder constructs `AvroGenericRecordReaderFunction`.

```
public outputType(Class<T> outputRecordTypeClass)
```

> we should also allow users to query metadata tables via the flink java api, right?

I am not sure if that is necessary. Doesn't Iceberg Java API cover that already?

Reply:
> > Can we provide a similar entry where the user simply tells the Builder that he wants to use AvroGenericRecordReaderFunction , and then lets the Builder create it?
> 
> I see your point. I agree that it is cumbersome for users to construct `AvroGenericRecordReaderFunction` or `RowDataReaderFunction`. Since we can extract type information from generic `T`, one idea would to add a setter for type class to the builder. If the type class is `RowData`, builder constructs `RowDataReaderFunction`. If it is Avro `GenericRecord, builder constructs `AvroGenericRecordReaderFunction`.
> 
> ```
> public outputType(Class<T> outputRecordTypeClass)
> ```
> 

I think this is a good idea, but I think this would merit another PR.
We should do this for the Source and the Sink as well.

> > we should also allow users to query metadata tables via the flink java api, right?
> 
> I am not sure if that is necessary. Doesn't Iceberg Java API cover that already?

For the metadata table, the most important use-case is when we read the metadata on the job initialisation and change the job based on the metadata read. Do we have use-case when we want to stream the results of the metadata query? Like when new files added to the table, or new snapshot created for the table?

IMHO, this is something for another PR if we decide to do it.

Reply:
@hililiwei I am merging this PR. please continue to comment on this thread on reader function. I will follow up with a separate PR on the `withOutputType` idea if we agree that is the right direction.

Reply:
> For the metadata table, the most important use-case is when we read the metadata on the job initialisation and change the job based on the metadata read. Do we have use-case when we want to stream the results of the metadata query? Like when new files added to the table, or new snapshot created for the table?

We have security, monitoring, and so on scenarios where these capabilities need to be used. Currently we implement this using `iceberg listeners`.